server/sonar-server/src/main/java/org/sonar/server/db/migrations/v52/MoveProjectProfileAssociation.java

```diff
@@ -0,0 +1,133 @@
+ 
+      if (profileKeysByLanguageThenName.contains(profileLanguage, profileName)) {
+        String profileKey = profileKeysByLanguageThenName.get(profileLanguage, profileName);
+
+        if (projectUuid == null) {
+          if (projectId == null || projectId == 0L) {
+            setDefaultProfile.setBoolean(1, true).setString(2, profileKey).execute();
+          } else {
+            LOGGER.warn(String.format("Profile with language '%s' and name '%s' is associated with unknown project '%d', ignored", profileLanguage, profileName, projectId));
+          }
+        } else {
+          associateProjectToProfile.setString(1, projectUuid).setString(2, profileKey).execute();
+        }
+      } else {
+        LOGGER.warn(String.format("Unable to find profile with language '%s' and name '%s', ignored", profileLanguage, profileName));
+      }
+
+      update.setLong(1, id);
+      return true;
+    }
+  }
+
+  private static final Logger LOGGER = Loggers.get(MoveProjectProfileAssociation.class);
+
+  public MoveProjectProfileAssociation(Database db) {
+    super(db);
+  }
+
+  @Override
+  public void execute(Context context) throws SQLException {
+
+    final Table<String, String, String> profileKeysByLanguageThenName = getProfileKeysByLanguageThenName(context);
+
+    MassUpdate massUpdate = context.prepareMassUpdate();
+    massUpdate.select("SELECT prop.id, prop.prop_key, prop.text_value, prop.resource_id, proj.uuid " +
+      "FROM properties prop " +
+      "LEFT OUTER JOIN projects proj ON prop.resource_id = proj.id " +
+      "WHERE prop.prop_key LIKE 'sonar.profile.%'"
+      );
+    massUpdate.update("DELETE FROM properties WHERE id = ?");
+
+    final Upsert setDefaultProfile = context.prepareUpsert("UPDATE rules_profiles SET is_default = ? WHERE kee = ?");
+    final Upsert associateProjectToProfile = context.prepareUpsert("INSERT INTO project_profiles (project_uuid, profile_key) VALUES (?, ?)");
+
+    try {
+      massUpdate.execute(new ProjectProfileAssociationHandler(setDefaultProfile, associateProjectToProfile, profileKeysByLanguageThenName));
+    } finally {
+      associateProjectToProfile.close();
+      setDefaultProfile.close();
+    }
+  }
+
+  private static String extractLanguage(String propertyKey) {
+    return propertyKey.substring("sonar.profile.".length());
+  }
+
+  private Table<String, String, String> getProfileKeysByLanguageThenName(final Context context) throws SQLException {
+    final Table<String, String, String> profilesByLanguageAndName = HashBasedTable.create();
+
+    Select selectProfiles = context.prepareSelect("SELECT kee, name, language FROM rules_profiles");
+    try {
+      selectProfiles.list(new RowReader<Void>() {
+        @Override
+        public Void read(Row row) throws SQLException {
+          profilesByLanguageAndName.put(row.getString(3), row.getString(2), row.getString(1));
+          return null;
+        }
+      });
+    } finally {
+      selectProfiles.close();
```

Comment:
Should not be needed as list() already calls close() when he has finished his job